centroid_decompostition.cpp

```diff
@@ -0,0 +1,92 @@
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+class Graph 
+{
+private:
+	int ver;
+	vector<int>* adj;
+	vector<int>* centroidTree;
+	int* sizes;
+	bool* deleted;
+public:
+	Graph(int num) : ver(num) {
+		adj = new vector<int>(ver + 1);
+		centroidTree = new vector<int>(ver + 1);
+		sizes = new int[ver + 1];
+		deleted = new bool[ver + 1];	
+	}
+	
+	~Graph();
+	
+	void addEdge(int x, int y, vector<int> arr[]);
+
+	void calcSizes(int cur, int par);
+
+	int findCentroid(int cur, int par, int vertices);
+
+	void decomposeTree(int cur, int total);
+};
+
+Graph::~Graph()
+{
+	delete []adj;
+	delete []sizes;
+	delete []deleted;
+}
+
+void Graph::addEdge(int x, int y, vector<int> arr[])
+{
+	arr[x].push_back(y);
+}
+
+void Graph::calcSizes(int cur, int par)
+{
+	sizes[cur] = 1;
+	for (int i = 0; i < adj[cur].size(); ++ i)
```

Comment:
```C++
for (const auto& element : adj[cur])
{
    if (!(elem == par || marked_[elem]))
    {
        calcSizes (elem, cur);
        sizes[cur] += sizes[elem]; // Do bounds checking here too
    }
}
```

Reply:
I'm not saying you shouldn't use it, but modern C++ has many benefits, and this is undoubtedly (At least to me) easier to read.